dds/DCPS/QueryConditionImpl.h

```diff
@@ -51,10 +51,19 @@ class OpenDDS_Dcps_Export QueryConditionImpl
 
   bool hasFilter() const;
 
+  /**
+   * Returns true if the sample matches the query.
+   */
   template<typename Sample>
-  bool filter(const Sample& s) const
+  bool filter(const Sample& s, bool has_invalid_data) const
   {
     ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex, guard, lock_, false);
+    const MetaStruct& meta = getMetaStruct<Sample>();
+    // Omit the sample from results if there are non-key fields in the query
+    // and the query only has key fields.
+    if (evaluator_.has_non_key_fields(meta) && has_invalid_data) {
```

Comment:
It seems like the call to "evaluator_.has_non_key_fields(meta)" is going to be constant for every pairwise combination of query string and sample type. Would it make sense to cache the result from this call somewhere? Possibly in a map inside FilterEvaluator?

Reply:
Once a QueryCondition is created, its expression can't change.  The QueryConditionImpl object can hold any additional info you need.